### PR TITLE
[WIP] HIVE-24445: Non blocking DROP table implementation

### PR DESCRIPTION
**Pull request by zchovan**
Change-Id: I594a724ccc6c836e6a735e9ef71851150b0f0c9b

### What changes were proposed in this pull request?

* added nem column to TBLS backend db table
* updated the related JDO entities
* added hive.txn.lockless.reads.enabled feature flag
* added writeId to DropTableOperation/Desc

### Why are the changes needed?

This is the first part of a bigger feature "lockless reads" the main goal is to avoid requesting locks for read operations, thus improving acid table performance

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

TBD

### CODE REVIEW

**Comment by zchovan** [comment]
@deniskuzZ

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableOperation.java [hunk 1]
```diff
@@ -104,6 +104,7 @@ public int execute() throws HiveException {
 
     // TODO: API w/catalog name
     context.getDb().dropTable(desc.getTableName(), desc.isPurge());
+//    context.getDb().dropTable(desc.getTableName(), desc.isPurge(), desc.getWriteId());
```
Comment (miklosgergely) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableOperation.java:107]: I assume that this was not intentionally left commented out.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableDesc.java [hunk 2]
```diff
@@ -29,31 +32,40 @@
...
-  public DropTableDesc(String tableName, boolean ifExists, boolean ifPurge, ReplicationSpec replicationSpec) {
-    this(tableName, ifExists, ifPurge, replicationSpec, true);
+  private long writeId = 0;
+
+  public DropTableDesc(TableName tableName, boolean ifExists, boolean ifPurge, ReplicationSpec replicationSpec) {
+    this(tableName, ifExists, ifPurge, replicationSpec, true, null);
   }
 
-  public DropTableDesc(String tableName, boolean ifExists, boolean purge, ReplicationSpec replicationSpec,
+  public DropTableDesc(TableName tableName, boolean ifExists, boolean ifPurge, ReplicationSpec replicationSpec,
```
Comment (miklosgergely) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableDesc.java:51]: Nit: the variable name should be purge - I know, it was ifPurge before, but it's a great opportunity to fix it :)

**Review by miklosgergely — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java [hunk 3]
```diff
@@ -54,9 +55,11 @@ public void analyzeInternal(ASTNode root) throws SemanticException {
       outputs.add(new WriteEntity(table, WriteEntity.WriteType.DDL_EXCLUSIVE));
     }
 
+    TableName tName = TableName.fromString(tableName, null, null);
+
     boolean purge = (root.getFirstChildWithType(HiveParser.KW_PURGE) != null);
     ReplicationSpec replicationSpec = new ReplicationSpec(root);
-    DropTableDesc desc = new DropTableDesc(tableName, ifExists, purge, replicationSpec);
+    DropTableDesc desc = new DropTableDesc(tName, ifExists, purge, replicationSpec);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableAnalyzer.java:62]: why is this change? i do not see constructor that accepts TableName. https://github.com/apache/hive/blob/08fd61d70fd172a63ee03e52f8c11db4718ba43c/ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/DropTableDesc.java#L41

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 4]
```diff
@@ -1259,6 +1261,43 @@ public boolean dropTable(String catName, String dbName, String tableName)
...
+      boolean success = false;
+      tableName = normalizeIdentifier(tableName);
+      dbName = normalizeIdentifier(dbName);
+      catName = normalizeIdentifier(catName);
+      Query query = null;
+
+      try {
+        LOG.debug("Executing markTableAsPendingDrop");
+
+        openTransaction();
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1277]: Why are you opening new transaction here?

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift [hunk 5]
```diff
@@ -300,6 +300,17 @@ struct TruncateTableRequest {
   5: optional string validWriteIdList
 }
 
+struct DropTableRequest {
+  1: required string dbName,
+  2: required string tableName,
+  3: required bool deleteDdata,
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:306]: what is it used for?

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift [hunk 6]
```diff
@@ -300,6 +300,17 @@ struct TruncateTableRequest {
...
+  4: optional i64 writeId=-1,
+  5: optional string validWriteIdList
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:308]: are you using this list anywhere?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 7]
```diff
@@ -1259,6 +1261,43 @@ public boolean dropTable(String catName, String dbName, String tableName)
     return success;
   }
 
+  private void markTableAsPendingDrop(String catName, String dbName, String tableName) {
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:1264]: Do you actually need to mark table as dropped or you could simply remove the record?

**Pull request closed without merging**